File: cpp/0054_Spiral_Matrix.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <queue>
#include <stack>
#include <deque>
#include <list>
#include <map>

using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        vector<vector<bool>> record(m, vector<bool>(n, false));
        vector<pair <int, int>> dir = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        int dir_idx = 0, x = 0, y = 0;
        vector<int> ans;

        while (ans.size() < m * n) {
            ans.push_back(matrix[x][y]);
            record[x][y] = true;

            int n_x = x + dir[dir_idx].first;
            int n_y = y + dir[dir_idx].second;

            if (n_x >= 0 && n_x < m && n_y >= 0 && n_y < n && !record[n_x][n_y]) {
                x = n_x;
                y = n_y;
            } else {
                dir_idx = (dir_idx + 1) % dir.size();
                x = x + dir[dir_idx].first;
                y = y + dir[dir_idx].second;
            }
        }

        return ans;
    }
};

int main(int argc, char *argv[]) {
    Solution solution;
    vector<vector<int>> q;
    vector<int> ans;

    // Input: matrix = [[1,2,3],[4,5,6],[7,8,9]]
    // Output: [1,2,3,6,9,8,7,4,5]
    q = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    ans = solution.spiralOrder(q);
    for (auto num : ans) {
        cout << num << " ";
    }
    cout << endl;

    // Input: matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
    // Output: [1,2,3,4,8,12,11,10,9,5,6,7]
    q = {{1, 2, 3, 4}, {5, 6, 7, 8}, {9, 10, 11, 12}};
    ans = solution.spiralOrder(q);
    for (auto num : ans) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}

File: py/0054_Spiral_Matrix.py
import sys
import heapq
from typing import *
from collections import OrderedDict
from collections import deque


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        ret = []
        m, n = len(matrix), len(matrix[0])
        visited = [[False for j in range(n)] for i in range(m)]
        delta = ((0, 1), (1, 0), (0, -1), (-1, 0))
        i, j, dir = 0, 0, 0

        while len(ret) != m * n:
            ret.append(matrix[i][j])
            visited[i][j] = True

            n_i, n_j = i + delta[dir][0], j + delta[dir][1]
            if n_i < 0 or n_i >= len(matrix) or n_j < 0 or n_j >= len(matrix[0]) or visited[n_i][n_j]:
                dir = (dir + 1) % 4
                i, j = i + delta[dir][0], j + delta[dir][1]
            else:
                i, j = n_i, n_j

        return ret


if __name__ == "__main__":
    solution = Solution()

    matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    print(solution.spiralOrder(matrix))

    matrix = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]
    print(solution.spiralOrder(matrix))

    sys.exit(0)
